contracts/truefi2/AaveBaseRateOracle.sol

```diff
@@ -0,0 +1,161 @@
+    function update() public offCooldown {
+        uint16 lidx = totalsBuffer.latestIndex;
+        uint16 nextIndex = (lidx + 1) % bufferSize();
+        uint256 apy = getAaveVariableBorrowAPY(asset);
+        uint256 dt = block.timestamp.sub(totalsBuffer.timestamps[lidx]);
+        totalsBuffer.runningTotals[nextIndex] = totalsBuffer.runningTotals[lidx].add(apy.add(totalsBuffer.lastValue).mul(dt).div(2));
+        totalsBuffer.timestamps[nextIndex] = block.timestamp;
+        totalsBuffer.lastValue = apy;
+        totalsBuffer.latestIndex = nextIndex;
+    }
+
+    /**
+     * @dev Average apy is calculated by taking
+     * the time-weighted average of the aave deposit apys.
+     * Essentially formula given below is used:
+     *
+     *           (v + v_{n-1}) / 2 * (t - t_{n-1}) + sum_{i=1}^{n - 1} (v_i + v_{i-1}) / 2 * (t_i - t_{i-1})
+     * avgAPY = -------------------------------------------------------------------------------------------
+     *                                                  (t - t_0)
+     *
+     * where v_i, t_i are values of the prices and their respective timestamps
+     * stored in the historical buffer, v is a value of current price ant t is its timestamp.
+     * Index n-1 corresponds to the most recent values and index 0 to the oldest ones.
+     *
+     * To avoid costly computations in a loop an optimization is used:
+     * Instead of directly storing aave deposit apys we store calculated numerators from formula above.
+     * This gives us most of the job done for every calculation.
+     *
+     * Notice that whether we are going to use the whole buffer or not
+     * depends on if it is filled up and the value of timeToCover parameter.
+     * @param timeToCover For how much time average should be calculated.
+     * @return Average apy.
+     */
+    function calculateAverageAPY(uint256 timeToCover) public view returns (uint256) {
+        require(
+            1 days <= timeToCover && timeToCover <= 365 days,
+            "AaveBaseRateOracle: Expected amount of time in range 1 to 365 days"
+        );
+        // estimate how much buffer we need to use
+        uint16 bufferSizeNeeded = uint16(timeToCover.div(cooldownTime));
+        require(bufferSizeNeeded <= bufferSize(), "AaveBaseRateOracle: Needed buffer size cannot exceed size limit");
+        uint16 lidx = totalsBuffer.latestIndex;
+        uint16 startIndex = (lidx + bufferSize() - bufferSizeNeeded + 1) % bufferSize();
+        if (totalsBuffer.timestamps[startIndex] == 0) {
+            startIndex = 0;
+        }
+        uint256 runningTotalForTimeToCover = totalsBuffer.runningTotals[lidx].sub(totalsBuffer.runningTotals[startIndex]);
+        uint256 curValue = getAaveVariableBorrowAPY(asset);
+        uint256 curTimestamp = block.timestamp;
+        uint256 dt = curTimestamp.sub(totalsBuffer.timestamps[lidx]);
+        runningTotalForTimeToCover = runningTotalForTimeToCover.add(curValue.add(totalsBuffer.lastValue).mul(dt).div(2));
```

Comment:
SPOT: Let's delete these lines so the average rate doesn't depend on a current spot rate. We'll sacrifice higher accuracy so that the average rate will be harder to game.